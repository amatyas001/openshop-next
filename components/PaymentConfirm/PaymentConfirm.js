import { Flex, Heading, SimpleGrid, Text } from '@chakra-ui/core';
import { useStripe, useElements } from '@stripe/react-stripe-js';
import { useDispatch } from 'react-redux';
import {
  paymentSuccess,
  paymentCancel,
  paymentError,
} from '../../redux/actions';
import { Button, CartContent } from '../index';

// Review and confirm payment intent
export const PaymentConfirm = ({ details, intent }) => {
  const stripe = useStripe();
  const elements = useElements();
  const dispatch = useDispatch();

  const handleConfirm = async () => {
    if (stripe) {
      const card = elements.getElement(CardElement);

      // confirm
      const result = await stripe.confirmCardPayment(intent.secret, {
        payment_method: {
          card: card,
          billing_details: {
            name: details.name,
            email: details.email,
            phone: details.phone,
          },
        },
      });

      // success
      if (result.paymentIntent && result.paymentIntent.status === 'succeeded') {
        dispatch(paymentSuccess(result.paymentIntent));
      }

      // error
      if (result.error) {
        dispatch(paymentError(result.error));
      }
    }
    return;
  };

  return (
    <Flex flexDirection='column' alignItems='center' width='100%'>
      {/* header */}
      <Heading>hey {details.name}!</Heading>
      <Text fontSize='1.1rem' fontWeight='bold'>
        Please review your payment details:
      </Text>

      {/* body */}
      <SimpleGrid columns={{ sm: 1, lg: 2 }} borderTop='1px' borderBottom='1px'>
        {/* intent */}
        <SimpleGrid
          columns='2'
          spacing='10px'
          fontSize='0.9rem'
          fontWeight='bold'
          flexDirection='column'
          p='3%'
        >
          <Text>INTENT ID</Text>
          <Text color='purple.800'>{intent.id}</Text>
          <Text>EMAIL</Text>
          <Text color='purple.800'>{details.email}</Text>
          <Text>PHONE</Text>
          <Text color='purple.800'>{details.phone}</Text>
          <Text>ADDRESS</Text>
          <Text color='purple.800'>{details.address}</Text>
        </SimpleGrid>

        {/* cart */}
        <CartContent />
      </SimpleGrid>

      {/* controls */}
      <SimpleGrid columns='2' spacing='15px'>
        {/* cancel */}
        <Button
          width='40%'
          bg='red.500'
          color='gray.100'
          onClick={() => {
            dispatch(paymentCancel(intent.id));
          }}
        >
          CANCEL ORDER
        </Button>

        {/* confirm */}
        <Button
          width='40%'
          bg='purple.800'
          color='gray.100'
          onClick={handleConfirm}
        >
          CONFIRM ORDER
        </Button>
      </SimpleGrid>
    </Flex>
  );
};
