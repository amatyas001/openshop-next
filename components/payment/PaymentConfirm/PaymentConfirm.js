import { useSelector } from 'react-redux';
import { loadStripe } from '@stripe/stripe-js';
import { Elements } from '@stripe/react-stripe-js';
import {
  CartContent,
  PaymentConfirmCard,
  PaymentConfirmControls,
  PaymentConfirmDetails,
  Spinner,
} from '@app/components';
import { Flex, Heading, Text } from '@chakra-ui/core';

const stripe = /* istanbul ignore next */ loadStripe(
  process.env.NEXT_PUBLIC_STRIPE_PUBLIC_KEY || 'test'
);

/**
 * Wrapper component for the confirm stage elements. See
 * [PaymentConfirmControls](#paymentconfirmcontrols) for
 * details about the next stages after *confirm* step.
 *
 * @visibleName Payment Confirm
 * @example
 * ```jsx
 * <PaymentConfirm />
 * ```
 */
export const PaymentConfirm = (props) => {
  const { details } = useSelector((store) => store.payment);
  const [loading, setLoading] = React.useState(false);
  const [handler, setHandler] = React.useState({
    complete: false,
    confirmHandler: /* istanbul ignore next */ () => {},
  });

  return (
    <>
      <Flex
        d={!loading ? 'flex' : 'none'}
        flexDirection='column'
        alignItems='center'
        width='100%'
        {...props}
      >
        <Heading data-testid='confirm-intent-name' my='1rem'>
          {`Hello, ${details.name}!`}
        </Heading>
        <Text fontSize='1.1rem' fontWeight='bold' my='0.7rem'>
          Please review your payment details and confirm order
        </Text>
        <PaymentConfirmDetails />
        <CartContent icons={false} width='100%' />
        <Elements stripe={stripe}>
          <PaymentConfirmCard loadHandler={setLoading} setHandler={setHandler} />
        </Elements>
        <PaymentConfirmControls
          confirmHandler={handler.confirmHandler}
          complete={handler.complete && !!details}
          loadHandler={setLoading}
        />
      </Flex>
      <Spinner d={loading ? 'flex' : 'none'} data-testid='confirm-spinner' text='Loading data...' />
    </>
  );
};
