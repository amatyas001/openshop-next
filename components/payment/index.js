export * from './PaymentCancelled/PaymentCancelled';
export * from './PaymentConfirm/PaymentConfirm';
export * from './PaymentConfirmCard/PaymentConfirmCard';
export * from './PaymentConfirmControls/PaymentConfirmControls';
export * from './PaymentConfirmDetails/PaymentConfirmDetails';
export * from './PaymentError/PaymentError';
export * from './PaymentForm/PaymentForm';
export * from './PaymentFormControls/PaymentFormControls';
export * from './PaymentProgress/PaymentProgress';
export * from './PaymentReview/PaymentReview';
export * from './PaymentSuccess/PaymentSuccess';
