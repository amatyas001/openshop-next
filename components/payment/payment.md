Each stage of the payment process is presented as child component under the `/checkout` page and assumes that `cart` is not empty. If `cart` is empty, than `/checkout` page will be redirect the user to the homepage.

Chekcout can be initiated from the [Cart](#section-cart) by clicking the _Continue_ button in the [Cart Panel](#cartpanel). This will load [PaymentReview](#paymentreview) stage and payment process becomes available under the `/checkout` page.

Components are interacting with the `payment` object and displayed according to `payment.status`. Payment data stored as various properties of the `payment` object. Detailed information can be found in each component documentation _(See: "State Dependencies")_.

Payment uses [Stripe](https://stripe.com) to handle secure customisable transactions. Requests to Stripe API are tokenized and _idempotent_.

On the case when user leave the payment process or close the browser tab, [Redux Persit](https://github.com/rt2zz/redux-persist) will store the current stage on the machine so process can be continued from where it was stopped.

**Stages**

1. **_Review_**

   - **[PaymentReview](#paymentreview)** is the initial stage. The user has options to go back browsing or continue the process.

2. **_Details_**

   1. **[PaymentForm](#paymentform)** is presented and on submit it will create an [Intent](https://stripe.com/docs/api/payment_intents/create) with [Client Secret](https://stripe.com/docs/payments/accept-a-payment).
   2. **[PaymentConfirm](#paymentconfirm)** using the _secret_ from [PaymentForm](#paymentform), asks the user to enter card details and confirm the payment.

3. **_Finished_**

   - **[PaymentSuccess](#paymentsuccess)** is shown when transaction was successful.
   - **[PaymentCancelled](#paymentcancelled)** is shown when user cancelled the payment process.
   - **[PaymentError](#paymenterror)** is shown when somthing bad happened during the transaction.

[PaymentProgress](#paymentprogress) is always visible on the `/checkout` page representing the payment process stages.

**State Dependencies** and **Wrapped Components** are documented in each description to give brief information about the context of posible interactions of the actual component. _Examples_ prepared with proper state and may render extra controls to test different cases with ease.
