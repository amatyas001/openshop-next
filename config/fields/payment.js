export const PAYMENT_FIELDS = [
  {
    name: 'name',
    type: 'text',
    placeholder: 'John Doe',
  },
  {
    name: 'address',
    type: 'text',
    placeholder: 'J191 S. Illinois Street, Everett, MA 02149',
  },
  ,
  {
    name: 'email',
    type: 'email',
    placeholder: 'johndoe@example.com',
  },
  {
    name: 'phone',
    type: 'tel',
    placeholder: '(941) 555-0123',
  },
];
