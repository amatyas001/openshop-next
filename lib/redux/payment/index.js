export * from './paymentCancel';
export * from './paymentCreate';
export * from './paymentStage';
