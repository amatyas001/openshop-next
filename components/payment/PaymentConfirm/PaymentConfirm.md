```jsx noeditor
import { useDispatch } from 'react-redux';
import { paymentConfirm } from '@app/lib/redux/actions';

const dispatch = useDispatch();

const details = {
  name: 'John Doe',
  email: 'johndoe@example.com',
  phone: '(122) 496 5737',
  address: 'J191 S. Illinois Street, Everett, MA 02149',
};

const intent = {
  id: 'intentid',
};

React.useEffect(() => {
  dispatch(paymentConfirm(details, intent));
}, []);

<>
  <PaymentConfirm />
</>;
```
