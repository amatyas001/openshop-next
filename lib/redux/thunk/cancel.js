import { paymentCancelled, paymentError } from '@app/redux/actions';

// Canceling intent on Stripe by intent ID (cancel param)
export function paymentCancel(cancel) {
  return async (dispatch) => {
    try {
      const axios = await import('axios');
      /* istanbul ignore next */
      const intent = await axios.post(
        process.env.NODE_ENV === 'production'
          ? 'https://openshop.netlify.app/.netlify/functions/intent'
          : 'http://localhost:9000/.netlify/functions/intent',
        {
          cancel,
        },
        {
          headers: { 'Content-Type': 'application/json' },
        }
      );

      // push to cancelled screen
      dispatch(paymentCancelled(intent.data));
    } catch (error) {
      // push to error screen
      dispatch(paymentError(error));
    }
  };
}
