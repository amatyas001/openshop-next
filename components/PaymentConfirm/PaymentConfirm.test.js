import { create, act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as thunks from '../../redux/thunk/cancel';
import * as actions from '../../redux/action/payment';
import { PaymentConfirm } from './PaymentConfirm';

const mockStore = configureStore([thunk]);

describe('<PaymentConfirm />', () => {
  let tree,
    store = mockStore({
      payment: { status: 'confirm', token: 'token' },
      cart: [{ id: 'item' }],
      amount: 100,
    }),
    intent = {
      id: 'mock_id',
      secret: 'secret',
    },
    details = {
      name: 'mock_name',
      email: 'mock_email',
      phone: 'mock_phone',
      address: 'mock_address',
    };

  it('should render without props', () => {
    act(() => {
      tree = create(
        <Provider store={store}>
          <PaymentConfirm />
        </Provider>
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  afterEach(() => {
    store = mockStore({
      payment: { status: 'confirm', token: 'token' },
      cart: [{ id: 'item' }],
      amount: 100,
    });
  });

  describe('with props', () => {
    beforeAll(() => {
      act(() => {
        tree = create(
          <Provider store={store}>
            <PaymentConfirm intent={intent} details={details} />
          </Provider>
        );
      });
    });

    it('should render intent id', () => {
      expect(
        tree.root.findByProps({ 'data-testid': 'confirm-intent-id' }).props
          .children
      ).toContain(intent.id);
    });

    for (let key in details) {
      it(`should render details ${key}`, () => {
        expect(
          tree.root.findByProps({ 'data-testid': `confirm-intent-${key}` })
            .props.children
        ).toContain(details[key]);
      });
    }

    it('should render cancel button', () => {
      expect(
        tree.root.findByProps({ 'data-testid': 'confirm-button-cancel' })
      ).toBeTruthy();
    });

    it('should render confirm button', () => {
      expect(
        tree.root.findByProps({ 'data-testid': 'confirm-button-confirm' })
      ).toBeTruthy();
    });

    it('should not display spinner', () => {
      try {
        tree.root.findByProps({ 'data-testid': 'confirm-spinner' });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  describe('on cancel', () => {
    let cancel;

    beforeAll(() => {
      cancel = jest
        .spyOn(thunks, 'paymentCancel')
        .mockImplementation(() => jest.fn());

      act(() => {
        tree.root
          .findByProps({ 'data-testid': 'confirm-button-cancel' })
          .props.onClick();
      });
    });

    it('should display spinner', () => {
      expect(
        tree.root.findByProps({ 'data-testid': 'confirm-spinner' })
      ).toBeTruthy();
    });

    it('should push cancel state', () => {
      expect(cancel).toHaveBeenCalledTimes(1);
      expect(cancel).toHaveBeenCalledWith(intent.id);
    });
  });

  describe('on confirm', () => {
    let success, error;

    beforeAll(async () => {
      success = jest
        .spyOn(actions, 'paymentSuccess')
        .mockImplementation(() => jest.fn());

      error = jest
        .spyOn(actions, 'paymentError')
        .mockImplementation(() => jest.fn());

      await act(async () => {
        await tree.root
          .findByProps({ 'data-testid': 'confirm-button-confirm' })
          .props.onClick();
      });
    });

    it('should display spinner', () => {
      expect(
        tree.root.findByProps({ 'data-testid': 'confirm-spinner' })
      ).toBeTruthy();
    });

    it('should push success state', async () => {
      expect(success).toHaveBeenCalledTimes(1);
    });

    it('should push error state', async () => {
      act(() => {
        tree = create(
          <Provider store={store}>
            <PaymentConfirm
              intent={{ ...intent, secret: 'error' }}
              details={details}
            />
          </Provider>
        );
      });

      await act(async () => {
        await tree.root
          .findByProps({ 'data-testid': 'confirm-button-confirm' })
          .props.onClick();
      });

      expect(error).toHaveBeenCalledTimes(1);
    });
  });
});
