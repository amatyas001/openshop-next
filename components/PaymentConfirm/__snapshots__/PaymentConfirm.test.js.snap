// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaymentConfirm /> should render with props 1`] = `
<div
  aria-hidden={false}
  className="css-1mi3tt8"
>
  <h2
    className="css-84al3w"
    data-testid="confirm-intent-name"
  >
    hey 
    mock_name
    !
  </h2>
  <p
    className="css-1358t0d"
  >
    Please review your payment details:
  </p>
  <div
    className="css-13cm5uv"
  >
    <div
      className="css-d34wob"
    >
      <p
        className="css-18poma1"
      >
        INTENT ID
      </p>
      <p
        className="css-1n4bukm"
        data-testid="confirm-intent-id"
      >
        mock_id
      </p>
      <p
        className="css-18poma1"
      >
        EMAIL
      </p>
      <p
        className="css-1n4bukm"
        data-testid="confirm-intent-email"
      >
        mock_email
      </p>
      <p
        className="css-18poma1"
      >
        PHONE
      </p>
      <p
        className="css-1n4bukm"
        data-testid="confirm-intent-phone"
      >
        mock_phone
      </p>
      <p
        className="css-18poma1"
      >
        ADDRESS
      </p>
      <p
        className="css-1n4bukm"
        data-testid="confirm-intent-address"
      >
        mock_address
      </p>
    </div>
    CartContent
  </div>
  <div
    className="css-dvxksz"
  >
    <button
      className="css-46aqdj"
      data-cy="button"
      data-testid="confirm-button-cancel"
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      CANCEL ORDER
    </button>
    <button
      className="css-187kagv"
      data-cy="button"
      data-testid="confirm-button-confirm"
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      CONFIRM ORDER
    </button>
  </div>
</div>
`;

exports[`<PaymentConfirm /> should render without props 1`] = `
<div
  aria-hidden={false}
  className="css-1mi3tt8"
>
  <h2
    className="css-84al3w"
    data-testid="confirm-intent-name"
  >
    hey 
    !
  </h2>
  <p
    className="css-1358t0d"
  >
    Please review your payment details:
  </p>
  <div
    className="css-13cm5uv"
  >
    <div
      className="css-d34wob"
    >
      <p
        className="css-18poma1"
      >
        INTENT ID
      </p>
      <p
        className="css-1n4bukm"
        data-testid="confirm-intent-id"
      />
      <p
        className="css-18poma1"
      >
        EMAIL
      </p>
      <p
        className="css-1n4bukm"
        data-testid="confirm-intent-email"
      />
      <p
        className="css-18poma1"
      >
        PHONE
      </p>
      <p
        className="css-1n4bukm"
        data-testid="confirm-intent-phone"
      />
      <p
        className="css-18poma1"
      >
        ADDRESS
      </p>
      <p
        className="css-1n4bukm"
        data-testid="confirm-intent-address"
      />
    </div>
    CartContent
  </div>
  <div
    className="css-dvxksz"
  >
    <button
      className="css-46aqdj"
      data-cy="button"
      data-testid="confirm-button-cancel"
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      CANCEL ORDER
    </button>
    <button
      className="css-187kagv"
      data-cy="button"
      data-testid="confirm-button-confirm"
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      CONFIRM ORDER
    </button>
  </div>
</div>
`;
